### PR TITLE
refactor: lake: manifest semver & code cleanup

### PR DESCRIPTION
**Pull request by tydeu**
Switches the manifest format to use `major.minor.patch` semantic versions. Major version increments indicate breaking changes (e.g., new required fields and semantic changes to existing fields). Minor version increments (after `0.x`) indicate backwards-compatible extensions (e.g., adding optional fields, removing fields). This change is backwards compatible. Lake will still successfully read old manifest with numeric versions. It will treat the numeric version `N` as semantic version `0.N.0`. Lake will also accept manifest versions with `-` suffixes (e.g., `x.y.z-foo`) and then ignore the suffix.

This change also includes the general cleanup/refactoring of the manifest code and data structures that was part of #3174.

### CODE REVIEW

File: src/lake/tests/manifest/lake-manifest-v1.0.0.json [hunk 1]
```diff
@@ -0,0 +1,20 @@
+{"version": "1.0.0",
```
Comment (Vtec234) [inline src/lake/tests/manifest/lake-manifest-v1.0.0.json:1]: Syntactic comment, but for somebody just looking at the manifest, this field is easy to confuse with the package version. But I suppose that was already the case while using sequential versioning.

**Comment by semorrison** [comment]
Just a minor comment about PR labelling: I think you might be overusing the "draft" status, or at least I don't think anything should ever be both a draft and `will-merge-soon`, which have opposite implications for reviewing urgency!

**Comment by tydeu** [comment]
> I don't think anything should ever be both a draft and will-merge-soon,

@semorrison That was just me forgetting to mark it ready for review. 🤦‍♂️

File: src/lake/tests/manifest/lake-manifest-v1.0.0.json (hunk 1 above)
Comment (tydeu) [reply src/lake/tests/manifest/lake-manifest-v1.0.0.json:1]: True, that is certainly something to keep in mind. I don't think I have good solution for that now in this PR, though.

**Pull request merged**